### PR TITLE
Support conversion of hexidecimal values

### PR DESCRIPTION
**Pull request by marcpaterno**
This small addition allows `as.integer64.character` to handle hexidecimal printed representation of large integers, in the style of `as.integer` in base R.

### CODE REVIEW

**Comment by MichaelChirico** [comment]
thanks for the PR! the idea seems great!

would you mind

 1. filing an issue that motivates the change, including the current failure mode / (optional) known workarounds 
 2. including a test that actually touches the 64-bit range, i.e. that wouldn't fit in 32-bit integers (or I guess that's covered already?, I never remember what the exact boundary is... maybe make it at least 2^33 for clarity)
 3. include a few more tests, e.g. of negative #s, hex #s that don't fit in 64 bits, any edge cases you can think of

**Comment by MichaelChirico** [comment]
WDYT about just letting `strtoll()` handle this by setting `base=0` instead?

https://en.cppreference.com/w/c/string/byte/strtol

File: src/integer64.c [hunk 1]
```diff
@@ -156,7 +156,17 @@ SEXP as_integer64_character(SEXP x_, SEXP ret_){
...
   for(i=0; i<n; i++){
     str = CHAR(STRING_ELT(x_, i)); endpointer = (char *)str; // thanks to Murray Stokely 28.1.2012
-    ret[i] = strtoll(str, &endpointer, 10);
+    int base = 10; // default
+    int sign = 1;  // default 
+    if (str[0]=='-'){
+      sign = -1;
+      str++;
+    }
+    if (str[0]=='0' && (str[1] == 'x' || str[1] == 'X')){
```
Comment (MichaelChirico) [inline src/integer64.c:165]: won't this be undefined behavior for `as.integer64('0')`? I guess 'no' because `str[0]=='0'` guarantees that `str[1]` is either `'\0'` or another string?

File: src/integer64.c [hunk 2]
```diff
@@ -156,7 +156,17 @@ SEXP as_integer64_character(SEXP x_, SEXP ret_){
   char * endpointer;
   for(i=0; i<n; i++){
     str = CHAR(STRING_ELT(x_, i)); endpointer = (char *)str; // thanks to Murray Stokely 28.1.2012
-    ret[i] = strtoll(str, &endpointer, 10);
+    int base = 10; // default
+    int sign = 1;  // default 
+    if (str[0]=='-'){
+      sign = -1;
+      str++;
```
Comment (MichaelChirico) [inline src/integer64.c:163]: `str` is `const` above, this will violate that, won't it?

**Comment by MichaelChirico** [comment]
I think the issue with just doing `base=0` is it will not be back-compatible (c.f. `as.integer64('011')`, which will be `11LL` on `main` but `9LL` with `base=0`).

We can easily imagine there's ppl working with long-string identifiers relying on the old behavior.

That said, I think we can do this?

```diff
-ret[i] = strtoll(str, &endpointer, 10);
+int negative=str[0] == '-';
+if (!strncmp(str+negative, "0x", 2) || !strncmp(str+negative,"0X", 2)) {
+  ret[i] = strtoll(str, &endpointer, 16);
+} else {
+  ret[i] = strtoll(str, &endpointer, 10);
+}
```

File: src/integer64.c (hunk 1 above)
Comment (marcpaterno) [reply src/integer64.c:165]: As you say, this is not undefined behavior. R guarantees the `str` is a pointer to a null-terminated character array. The test of `str[0]` means that the character at `str[1]` is either NUL or some other character, and testing it is valid.

I will extends the tests to verify.

**Comment by marcpaterno** [comment]
> I think the issue with just doing `base=0` is it will not be back-compatible (c.f. `as.integer64('011')`, which will be `11LL` on `main` but `9LL` with `base=0`).
> 
> We can easily imagine there's ppl working with long-string identifiers relying on the old behavior.
> 
> That said, I think we can do this?
> 
> ```diff
> -ret[i] = strtoll(str, &endpointer, 10);
> +int negative=str[0] == '-';
> +if (!strncmp(str+negative, "0x", 2) || !strncmp(str+negative,"0X", 2)) {
> +  ret[i] = strtoll(str, &endpointer, 16);
> +} else {
> +  ret[i] = strtoll(str, &endpointer, 10);
> +}
> ```

I have tested this code and it yields different behavior in one edge case. I thought I had picked the appropriate behavior, but you can of course correct me.

In the case of handling the string "-0x8000000000000000", my original code yielded -9223372036854775807, the first value in the result of `lim.integer64()`. Modified as you propose above, the result is "NA".

I will add the new testing (that includes the handling of input "0") to the PR, and await your decision on the desired behavior for handling "-0x8000000000000000".

File: src/integer64.c (hunk 2 above)
Comment (marcpaterno) [reply src/integer64.c:163]: If `str` is a pointer-to-const, this does not violate the const-ness. We are not writing through the pointer to modify `*str`; we are only incrementing our copy of `str`, which does not modify the value in the caller.

Comment (MichaelChirico) [reply src/integer64.c:163]: oh right, basically it's `(const str) *`, not `const (str *)`

**Comment by MichaelChirico** [comment]
Let's pick whatever `strtoll("-0x8000000000000000", &endpointer, 16)` does, I think it's best to be consistent with & as close to the well-tested underlying C library routine.

**Comment by MichaelChirico** [comment]
please add a NEWS entry btw!

File: tests/testthat/test-integer64.R [hunk 3]
```diff
@@ -8,9 +8,7 @@ test_that("integer64 coercion to/from other types work", {
 
   # to integer64
   expect_identical(as.integer64(TRUE), as.integer64(1L))
-  expect_identical(as.integer64(as.character(1:10)), as.integer64(1:10))
-  expect_identical(as.integer64(as.double(1:10)), as.integer64(1:10))
-  expect_identical(as.integer64(NULL), as.integer64())
+  expect_identical(as.integer64(as.character(-10:10)), as.integer64(-10:10))
```
Comment (MichaelChirico) [inline tests/testthat/test-integer64.R:11]:
```suggestion
  expect_identical(as.integer64(as.character(-10:10)), as.integer64(-10:10))
  expect_identical(as.integer64(as.double(-10:10)), as.integer64(-10:10))
  expect_identical(as.integer64(NULL), as.integer64())
```

**Comment by marcpaterno** [comment]
I do not think it is possible to match the behavior of `strtoll` with the current representation of `integer64`.
`strtoll` will read "-9223372036854775808" and represent it exactly. This value is 1 smaller than what `lim.integer64()` indicates can be represented by `integer64`. `strtoll` will read "-9223372036854775809" and yield the value -9223372036854775808.

How would you like to proceed?

**Comment by MichaelChirico** [comment]
Oh, great point. That's because -922...808 is how `NA_integer64_` is represented (just as `NA_integer_` is represented as -214...647):

https://github.com/r-lib/bit64/blob/62cd4ee84356224feb939632b5d04c802bb72306/src/integer64.h#L28

In fact you've found what I think is kind of buggy behavior in `as.integer64.character` for decimal string input, too: #175.

Part of the issue is, there's no 128-bit double to case into with full fidelity to work with these out-of-range values before casting back down to 64 bits; that's what's done in the 32-bit case by R:

https://github.com/r-devel/r-svn/blob/c8ab68478c5d2322956e19764c20c092916133cd/src/main/util.c#L2112

Anyway, for now, I think we should aim to be consistent between hex & decimal strings, and to bitstrings & return `NA_integer64_` for `-9223372036854775808`

```r
# as.bitstring(lim.integer64())
as.integer64(structure("1000000000000000000000000000000000000000000000000000000000000000", class='bitstring'))
# integer64
# [1] <NA>
```
